### ISSUE TITLE
It looks like LZ forgot how to pass

### ISSUE THREAD

**Issue by afalturki** [leela-zero/leela-zero#1446]
Before ELF self-play games were introduced, LZ was great at passing correctly when the game was finished and all captured stones were removed from the board. However, after ELF self-play games were introduced, it seems like LZ is much worse now at passing and the average number of moves for no-resign self-play games became much higher.

Below are some examples:

http://zero.sjeng.org/view/285b34be55e246beb28c6f0196af18eca78c0c577c319413c024cd4d6c58bd40?viewer=wgo

http://zero.sjeng.org/view/b6d8622b205a166d8f86414f848316ed87c4a7b35baff5e7a846751ee64d9106?viewer=wgo

http://zero.sjeng.org/view/d541a93df4088635d52097d7eaa8f75aaf474e75e20e457560eea2677d3c7e2d?viewer=wgo

http://zero.sjeng.org/view/6435b0c7a388a0233ebfffca0ab4722c837f97d80196d56b4ba9390ecbc1768d?viewer=wgo

http://zero.sjeng.org/view/2bff8b453c590d54d2f5fc77916caa0e2a1fe718d2e4ba8af78795a578e9b3b2?viewer=wgo

This is a no-resign game that's working normally though:
http://zero.sjeng.org/view/045b5f38524b3770c1430f8d00333a6b911505f430058f1bcf8972856a58d265?viewer=wgo

**Comment by jkiliani** [comment]
I'm not really surprised there are drawbacks to using self-play from a network that does not get reinforcement learning feedback itself. I'd suggest we should stop training on the no-resign ELF games and only use the newer ones with resignation enabled for training. Eventually, reinforcement learning should teach LZ nets to pass again in self-play, once there is no longer a training impulse in the opposite direction.

**Comment by lp200** [comment]
I think that it is effect of -m 999

**Comment by Hersmunch** [comment]
Has anyone run the resign analysis lately?
I'll turn on the necessary output on my autogtp now but my gpu isn't super fast so it'll be a while before I have some results..

**Comment by afalturki** [comment]
@jkiliani I think all self-play ELF games are resign-enabled.

**Comment by Mardak** [comment]
> I think that it is effect of -m 999

I believe this is right. T=1 makes it harder to pick "pass" from multiple moves that are all 100% win rate.

Looking closer at http://zero.sjeng.org/view/285b34be55e246beb28c6f0196af18eca78c0c577c319413c024cd4d6c58bd40?viewer=wgo

The first white pass is move 222, so the first opportunity for black to pass i 223. Turns out pre-ELF ecab83bb, current 192x15 LZ139 a39b64cc, and 224x20 all don't want to pass yet. (All have average prior for passing around 0.3%.)

The second white pass is move 444, and here's the moves it's trying to select from:
```
expect: pass, move: 445, who: b, visit: 1000, symmetry: x̄
weight: ecab83bb
 A18 ->      35 (V: 100.00%) (N:  3.52%) PV: A18 C19 D19 B18 A17 B17
  S4 ->      28 (V: 100.00%) (N:  2.74%) PV: S4 T4 T3 J2 T5
  H2 ->      27 (V: 100.00%) (N:  2.73%) PV: H2 pass A18 B18 A17
 M19 ->      27 (V: 100.00%) (N:  2.68%) PV: M19 pass A18 pass B18
  R2 ->      26 (V: 100.00%) (N:  2.57%) PV: R2 J2 H2 pass A18
 T11 ->      25 (V: 100.00%) (N:  2.49%) PV: T11 T19 T18 S18 T19 T17 S17 T16 S16
 E16 ->      25 (V: 100.00%) (N:  2.48%) PV: E16 pass A18 C19 D19
  K9 ->      24 (V: 100.00%) (N:  2.43%) PV: K9 J2 H2 pass A18
 K19 ->      24 (V: 100.00%) (N:  2.43%) PV: K19 J2 H2 pass A18
 G14 ->      24 (V: 100.00%) (N:  2.42%) PV: G14 pass A18 pass A17
 H14 ->      24 (V: 100.00%) (N:  2.42%) PV: H14 pass A18 pass B18
 A14 ->      24 (V: 100.00%) (N:  2.41%) PV: A14 pass A18 pass A16
  P1 ->      24 (V: 100.00%) (N:  2.40%) PV: P1 pass A18 pass B18
  J2 ->      24 (V: 100.00%) (N:  2.37%) PV: J2 pass A18 pass B18
 H18 ->      24 (V: 100.00%) (N:  2.36%) PV: H18 pass A18 B18 A17
 G12 ->      23 (V: 100.00%) (N:  2.34%) PV: G12 pass A18 pass A17
 H13 ->      23 (V: 100.00%) (N:  2.33%) PV: H13 pass A18 pass A17
  K4 ->      23 (V: 100.00%) (N:  2.33%) PV: K4 L5 A18 C19 D19 J2 B19 C18
 J11 ->      23 (V: 100.00%) (N:  2.32%) PV: J11 pass A18 pass A17
 O15 ->      23 (V: 100.00%) (N:  2.30%) PV: O15 pass A18 C19 D19
 K17 ->      23 (V: 100.00%) (N:  2.30%) PV: K17 pass A18 C19 D19
 K14 ->      23 (V: 100.00%) (N:  2.28%) PV: K14 pass A18 C19 D19
  F1 ->      23 (V: 100.00%) (N:  2.27%) PV: F1 pass A18 pass B18
 J15 ->      22 (V: 100.00%) (N:  2.24%) PV: J15 J2 H2 pass A18
  O2 ->      22 (V: 100.00%) (N:  2.21%) PV: O2 O4 A18 C19 D19
  L3 ->      22 (V: 100.00%) (N:  2.21%) PV: L3 M1 N1 J2 H2
 K18 ->      22 (V: 100.00%) (N:  2.21%) PV: K18 pass A18 pass B18
 M14 ->      22 (V: 100.00%) (N:  2.20%) PV: M14 pass A18 pass B18
  E1 ->      22 (V: 100.00%) (N:  2.19%) PV: E1 pass A18 pass B18
 H17 ->      22 (V: 100.00%) (N:  2.16%) PV: H17 pass A18 pass B18
 H19 ->      22 (V: 100.00%) (N:  2.16%) PV: H19 J2 H2 pass A18
 B11 ->      21 (V: 100.00%) (N:  2.13%) PV: B11 pass A18 pass B18
 A12 ->      21 (V: 100.00%) (N:  2.12%) PV: A12 pass A18 pass B18
 P17 ->      21 (V: 100.00%) (N:  2.09%) PV: P17 pass A18 pass B18
 G18 ->      21 (V: 100.00%) (N:  2.06%) PV: G18 pass A18 C19 D19
  E4 ->      20 (V: 100.00%) (N:  2.03%) PV: E4 pass A18 pass B18
  R5 ->      20 (V: 100.00%) (N:  2.02%) PV: R5 pass A18 pass B18
  E2 ->      20 (V: 100.00%) (N:  2.02%) PV: E2 pass A18 pass B18
  B9 ->      20 (V: 100.00%) (N:  2.01%) PV: B9 pass A18 pass A17
 B13 ->      20 (V: 100.00%) (N:  1.97%) PV: B13 pass A18 pass B18
  C3 ->      19 (V: 100.00%) (N:  1.91%) PV: C3 pass A18 pass B18
pass ->      19 (V: 100.00%) (N:  1.90%) PV: pass 
  T1 ->      16 (V: 100.00%) (N:  1.65%) PV: T1 J2 H2 R2 S2
  Q3 ->      16 (V: 100.00%) (N:  1.59%) PV: Q3 pass A18 pass B18

weight: 192x15
 A18 ->      30 (V: 100.00%) (N:  3.03%) PV: A18 B18 A17 B17 B19 C19
  P1 ->      26 (V: 100.00%) (N:  2.59%) PV: P1 H18 H19 H17 A18
  S4 ->      26 (V: 100.00%) (N:  2.55%) PV: S4 T3 T4 H18 S3 H17
 M19 ->      25 (V: 100.00%) (N:  2.54%) PV: M19 H18 H19 H17 A18
  H2 ->      25 (V: 100.00%) (N:  2.52%) PV: H2 H17 A18 B18 A17
 H13 ->      24 (V: 100.00%) (N:  2.37%) PV: H13 H18 H19 H17 A18
 A14 ->      24 (V: 100.00%) (N:  2.36%) PV: A14 H2 A18 B18 A17
pass ->      23 (V: 100.00%) (N:  2.35%) PV: pass 
 T11 ->      23 (V: 100.00%) (N:  2.34%) PV: T11 T19 T18 S18 T19 T17 S17 R17 A18
 E16 ->      23 (V: 100.00%) (N:  2.34%) PV: E16 H2 A18 B18 A17
 H19 ->      23 (V: 100.00%) (N:  2.33%) PV: H19 H18 A18 B18 A17
  E1 ->      23 (V: 100.00%) (N:  2.32%) PV: E1 H17 A18 B18
 H18 ->      23 (V: 100.00%) (N:  2.32%) PV: H18 G14 A18 B18 A17
  J2 ->      23 (V: 100.00%) (N:  2.30%) PV: J2 pass A18 B18 A17
 K19 ->      23 (V: 100.00%) (N:  2.30%) PV: K19 H18 H19 H17 A18
 J15 ->      23 (V: 100.00%) (N:  2.30%) PV: J15 H18 H19 H17 A18
 H17 ->      23 (V: 100.00%) (N:  2.29%) PV: H17 H18 H19 K18 K19
 H14 ->      23 (V: 100.00%) (N:  2.29%) PV: H14 H17 A18 B18 A17
  O2 ->      23 (V: 100.00%) (N:  2.29%) PV: O2 O4 A18 A17 A16
 G14 ->      23 (V: 100.00%) (N:  2.29%) PV: G14 H17 A18 B18 A17
  F1 ->      23 (V: 100.00%) (N:  2.28%) PV: F1 H18 H19 H17 A18
 G12 ->      23 (V: 100.00%) (N:  2.28%) PV: G12 H18 H19 H17 A18
  K9 ->      23 (V: 100.00%) (N:  2.28%) PV: K9 H18 H19 H17 A18
 K17 ->      23 (V: 100.00%) (N:  2.27%) PV: K17 H18 H19 H17 A18
  K4 ->      23 (V: 100.00%) (N:  2.27%) PV: K4 H2 J2 H1 H2 L5 A18 C19
 M14 ->      23 (V: 100.00%) (N:  2.27%) PV: M14 H18 H19 H17 A18
  R2 ->      23 (V: 100.00%) (N:  2.26%) PV: R2 H18 H19 H17 A18
  L3 ->      23 (V: 100.00%) (N:  2.26%) PV: L3 M1 N1 M2 A18 A17
  T1 ->      22 (V: 100.00%) (N:  2.25%) PV: T1 H2 A18 B18 A17
  E4 ->      22 (V: 100.00%) (N:  2.25%) PV: E4 H18 H19 H17 A18
 K14 ->      22 (V: 100.00%) (N:  2.24%) PV: K14 H18 H19 H17 A18
 J11 ->      22 (V: 100.00%) (N:  2.21%) PV: J11 H18 H19 H17 A18
 O15 ->      22 (V: 100.00%) (N:  2.18%) PV: O15 H18 H19 H17 A18
 P17 ->      22 (V: 100.00%) (N:  2.18%) PV: P17 H18 H19 H17 A18
  E2 ->      22 (V: 100.00%) (N:  2.16%) PV: E2 H18 A18 B18 A17
 A12 ->      21 (V: 100.00%) (N:  2.14%) PV: A12 H2 A18 B18 A17
 K18 ->      21 (V: 100.00%) (N:  2.11%) PV: K18 H17 A18 B18 A17
 B13 ->      21 (V: 100.00%) (N:  2.10%) PV: B13 H2 A18 B18 A17
 G18 ->      21 (V: 100.00%) (N:  2.09%) PV: G18 H2 A18 B18 A17
  C3 ->      21 (V: 100.00%) (N:  2.08%) PV: C3 H18 H19 H17 A18
  R5 ->      20 (V: 100.00%) (N:  2.03%) PV: R5 H2 A18 B18 A17
 B11 ->      20 (V: 100.00%) (N:  2.02%) PV: B11 H18 H19 H17 A18
  B9 ->      19 (V: 100.00%) (N:  1.96%) PV: B9 H18 H19 H17 A18
  Q3 ->      18 (V: 100.00%) (N:  1.83%) PV: Q3 H2 A18 B18 A17

weight: 224x20
  O2 ->      29 (V: 99.90%) (N:  2.84%) PV: O2 O3 S4 T4 pass
pass ->      27 (V: 100.00%) (N:  2.58%) PV: pass 
  L3 ->      27 (V: 99.90%) (N:  2.66%) PV: L3 M1 O2 M2 S4
  S4 ->      27 (V: 99.90%) (N:  2.68%) PV: S4 K18 pass K19 pass
  K4 ->      26 (V: 99.92%) (N:  2.55%) PV: K4 H2 A18 B18 L3 L1 pass B19
 A18 ->      26 (V: 99.90%) (N:  2.63%) PV: A18 B19 O2 K18 S4 T4
  H2 ->      26 (V: 99.89%) (N:  2.60%) PV: H2 K18 O2 K19 S4
 B11 ->      25 (V: 99.89%) (N:  2.50%) PV: B11 K18 O2 K19 S4
  E4 ->      24 (V: 99.89%) (N:  2.39%) PV: E4 J2 O2 F1 S4
 G14 ->      24 (V: 99.89%) (N:  2.40%) PV: G14 K18 O2 K19 S4
  B9 ->      24 (V: 99.89%) (N:  2.37%) PV: B9 J2 O2 F1 S4
  K9 ->      24 (V: 99.89%) (N:  2.38%) PV: K9 K18 O2 K19 S4
  R2 ->      23 (V: 99.90%) (N:  2.27%) PV: R2 K18 O2 K19 S4
  C3 ->      23 (V: 99.89%) (N:  2.33%) PV: C3 J2 O2 F1 S4
 H13 ->      23 (V: 99.89%) (N:  2.34%) PV: H13 K18 O2 K19 S4
 O15 ->      23 (V: 99.89%) (N:  2.31%) PV: O15 K18 O2 K19 S4
 M19 ->      23 (V: 99.88%) (N:  2.33%) PV: M19 K18 O2 O3 S4
 J11 ->      22 (V: 99.90%) (N:  2.18%) PV: J11 K18 O2 K19 S4
 K17 ->      22 (V: 99.89%) (N:  2.21%) PV: K17 K18 O2 H18 pass
 H17 ->      22 (V: 99.89%) (N:  2.24%) PV: H17 K18 O2 H18 S4
 E16 ->      22 (V: 99.89%) (N:  2.17%) PV: E16 K18 O2 K19 S4
  R5 ->      22 (V: 99.89%) (N:  2.20%) PV: R5 K18 O2 K19 S4
 B13 ->      22 (V: 99.89%) (N:  2.17%) PV: B13 K18 O2 K19 S4
 A14 ->      22 (V: 99.88%) (N:  2.25%) PV: A14 K18 O2 K19 S4
  J2 ->      22 (V: 99.88%) (N:  2.24%) PV: J2 K18 O2 K19 S4
 K18 ->      22 (V: 99.88%) (N:  2.26%) PV: K18 H18 O2 O4
 K19 ->      22 (V: 99.88%) (N:  2.24%) PV: K19 J2 O2 F1 S4
 G18 ->      22 (V: 99.88%) (N:  2.21%) PV: G18 K18 O2 O3 S4
  P1 ->      22 (V: 99.88%) (N:  2.17%) PV: P1 K18 O2 K19 S4
 H14 ->      22 (V: 99.88%) (N:  2.19%) PV: H14 H18 O2 K18
  T1 ->      22 (V: 99.88%) (N:  2.17%) PV: T1 K18 O2 K19 S4
 T11 ->      21 (V: 99.92%) (N:  2.11%) PV: T11 T19 T18 S17 T19 S16 S4 T4 T3
  Q3 ->      21 (V: 99.89%) (N:  2.11%) PV: Q3 K18 O2 O3 pass
  E1 ->      21 (V: 99.89%) (N:  2.07%) PV: E1 K18 O2 K19
 A12 ->      21 (V: 99.89%) (N:  2.12%) PV: A12 K18 O2 K19 S4
 H18 ->      21 (V: 99.88%) (N:  2.13%) PV: H18 K18 O2 O3
 H19 ->      21 (V: 99.88%) (N:  2.11%) PV: H19 K18 O2 K19 S4
 J15 ->      21 (V: 99.88%) (N:  2.13%) PV: J15 K18 O2 K19 S4
 P17 ->      21 (V: 99.88%) (N:  2.16%) PV: P17 K18 O2 K19 S4
 G12 ->      20 (V: 99.89%) (N:  2.05%) PV: G12 K18 O2 K19 S4
  E2 ->      20 (V: 99.89%) (N:  2.02%) PV: E2 F1 O2 H2 K4
  F1 ->      20 (V: 99.89%) (N:  2.02%) PV: F1 K18 O2 K19 S4
 K14 ->      20 (V: 99.88%) (N:  1.99%) PV: K14 K18 O2 H18 S4
 M14 ->      19 (V: 99.89%) (N:  1.91%) PV: M14 K18 O2 K19 S4
```

Notice that "pass" does exist somewhere in each of those moves lists. It's just that with T=1, the pass move will be picked about 2% of the time as everything looks good. Although with T=0, none of these networks would have picked pass either.. although ELF was just 2 visits off.

**Comment by alreadydone** [comment]
Scaling factor wr(1-wr) for c_puct as @jkiliani suggested together with precision fix #1440 may help with this.

**Comment by Mardak** [comment]
I wonder if it's worthwhile to have something like smarter passing instead of --dumbpass. --slightlysmarterpass ? ;) I don't know if it's the `wr(1-wr) for c_puct` or maybe just `if winrate == 100% && child.get_move() == PASS` in `uct_select_child` ?? Just to avoid continuing games unnecessarily.

**Comment by lp200** [comment]
Probably -m 400 would be a good choice.

**Comment by Marcin1960** [comment]
@lp200 "-m 400 would be"

What is this mysterious "-m"?

**Comment by jkiliani** [comment]
@Marcin1960 m = --randoMcnt, i.e. number of moves with temperature = 1.

**Comment by Marcin1960** [comment]
@jkiliani 

You mean this line in Leela.cpp?

        ("randomcnt,m", po::value<int>()->default_value(cfg_random_cnt),
                        "Play more randomly the first x moves.")

**Comment by bood** [comment]
I don't see a reason why Black in this game keep filling its own space in the end game:
http://www.yss-aya.com/cgos/viewer.cgi?19x19/SGF/2018/05/18/426628.sgf

Not sure if it's a setting issue or network issue.

**Comment by Mardak** [comment]
Do people have examples of where ELF doesn't know to pass? I was checking https://github.com/gcp/leela-zero/issues/1451#issuecomment-390245896 and both games, ELF passes just fine (assuming T=0):
```
http://www.yss-aya.com/cgos/viewer.cgi?19x19/SGF/2018/05/18/426612.sgf
expect: pass, move: 270, who: w, weight: 224x20, visit: 3200
pass ->    1650 (V: 100.00%) (N: 35.19%) PV: pass 
 F18 ->      63 (V: 99.93%) (N:  1.45%) PV: F18 A19 pass C19 pass B17 pass A18 pass
  A8 ->      61 (V: 99.91%) (N:  1.43%) PV: A8 A18 pass B19 pass C18 pass B17
 T19 ->      56 (V: 99.91%) (N:  1.31%) PV: T19 A18 pass B19 pass C18 pass B17
…

http://www.yss-aya.com/cgos/viewer.cgi?19x19/SGF/2018/05/18/426628.sgf
expect: pass, move: 285, who: b, weight: 224x20, visit: 3200
pass ->     883 (V: 100.00%) (N: 21.21%) PV: pass 
  J2 ->     348 (V: 99.94%) (N:  8.82%) PV: J2 pass pass
  J1 ->     181 (V: 99.93%) (N:  4.63%) PV: J1 pass pass
  S6 ->     149 (V: 99.91%) (N:  3.88%) PV: S6 pass pass
…
```

**Comment by alreadydone** [comment]
@bood Interesting, it fills up all possible spaces it can from left to right, from top to bottom! Looks like winrate is identically 100%, and it may have used a winrate(1-winrate) factor for c_puct so that the policy also can't distinguish between moves, so genmove selects every move that does not lose in a particular order. Since "pass" is the last move in the order, it will never pass until all possible spaces are filled.

**Comment by Mardak** [comment]
From replaying the sgf, it seems unlikely that all the moves had the same visits with -p 1600 (but they probably do all have the same eval of 100%). Even if actual visit counts are being cut short with time management, the priors for pass seem to be multiple times higher than other moves, so it should get more visits…

Anyone know who is running the bot and if there's logs?

**Comment by jokkebk** [comment]
@alreadydone yes and it is also VERY annoying if you watch the game with LeelaWatcher, to see for example white ahead by 200 points, black having killed all his groups but one which has 1 eye and no liberties, and game continues for half an hour with black inserting stones in empty areas of white territory (usually at this point there are like 200 white stones and 100 black ones, so essentially filling small gaps with dead stones), and white is ignoring the +200 point capture and instead fills his own liberties one by one, maybe passing every 70 moves or so for a change.

Doesn't maybe hurt the training process, but while my GTX 1060 is wheezing at full power to generate this "training", it feels like there should be a red "RESIGN FOR GODSAKES" -button on LeelaWatcher to help black out of his misery and get on with some useful training. 🤣

**Comment by alreadydone** [comment]
I compiled Windows binaries with c_puct scaling (see https://github.com/gcp/leela-zero/pull/1435). Using the linear version, I can reproduce 59% of the moves of @bood's game between move 315 and move 393, but there are many passes among the non-reproduced moves, so even a lucky random seed is unlikely to keep playing for so long. Using Lizzie, I did observe that moves pop out starting from the top left corner. The quadratic version always passes since move 277 (contrary to my thoughts above!).

Another thing I observed is that black winrate drops from 100% a bit (up to 2+ percent) when the board gets crowded.
